Add more file extensions

robertodr opened the pull request:
I am not a Ruby expert, but I've managed to make the following changes:
- [x] The plugin now also checks files with extensions `.hpp`, `.hh`, `.cxx`, `.cc`, `.cpp`, hence addressing #2  
- [x] The error message is more informative. It first lists the offending files, then suggests appropriate actions and eventually prints the patches to apply.
- [x] Generation of the more informative error message is now done differently. 
- [x] Pass the list of extensions to check to the plugin, instead of hardcoding the valid extensions. Default behavior only checks Objective-C extensions (`.h`, `.m`, `.mm`)

You can see the plugin in action for this example repository: https://github.com/PCMSolver/danger-cpp-example/pull/3

~### Questions/Ideas~

~- [ ] I haven't added tests, because I don't know how to.~
~- [ ]Generating the more informative error message requires iterating twice over the `changes` array.~

nikolaykasyanov commented:
Looks great!

I believe it's a good start, a more generic approach (passing extensions as argument) can be implemented later.

Adding/editing tests is easy, most of them are [here](https://github.com/flix-tech/danger-code_style_validation/blob/a9dc5f4a37e4684db7f193e84401dc65c64ba859/spec/code_style_validation_spec.rb).

You might also want to run tests locally, to do that you need to install bundler gem and then do `bundle install` in the project directory. When it's done, run `bundle exec rake spec` to run tests.

When in doubt, please use to [.travis.yml](https://github.com/flix-tech/danger-code_style_validation/blob/a9dc5f4a37e4684db7f193e84401dc65c64ba859/.travis.yml) file as a reference.

Thank you for your contribution and feel free to ask anything.

robertodr commented:
Thank for the pointers, @nikolaykasyanov. I have now fixed the existing tests and refactored the generation of the error message. I will try to add a tests for:
1. More than one failing file.
2. The new file extensions.
If you have suggestions on how to pass file extensions to be checked to the plugin, I can try to implement that too.

nikolaykasyanov commented:
@robertodr regarding file extension list as argument:

we can do that by introducing a new key into plugin's configuration hash. At the time only [`ignore_file_patterns`](https://github.com/flix-tech/danger-code_style_validation/blob/a9dc5f4a37e4684db7f193e84401dc65c64ba859/lib/code_style_validation/plugin.rb#L27) exists.

A new key could be introduced and its value could be used instead of [hardcoded list of extensions](https://github.com/flix-tech/danger-code_style_validation/blob/a9dc5f4a37e4684db7f193e84401dc65c64ba859/lib/code_style_validation/plugin.rb#L72).

Also, if this key is not specified in the config, current list of extensions must be used for backward compatibility reasons.

Does it make sense?

robertodr commented:
How does it look now? :smile: I modified one line in the tests, so that the new parameter is tested:
```ruby
@my_plugin.check file_extensions: ['.h', '.m', '.mm', '.C', '.cpp']
```
Adding a new proper test for a `.hpp` and/or `.cpp` file is beyond my capabilities, however.

nikolaykasyanov commented on README.md:
```diff
@@ -2,9 +2,10 @@
+This plugin uses 'clang-format' to look for code style violations in added
+lines on the current MR / PR, and offers inline patches.
+By default only Objective-C files, with extensions `.h`, `.m`, `.mm` and
+`.C`, are checked.
```
Why does `.C` have to be capitalized?

robertodr replied:
According to [Wikipedia](https://en.wikipedia.org/wiki/Objective-C) Objective-C files usually have those extensions.

nikolaykasyanov replied:
Good find, I've never seen that extension in the wild, though. Also, it's effectively equal to `.c` extension on case-insensitive filesystems (i.e. on most Macs).

robertodr replied:
Yes, of course! I'll remove it and leave `.m`, `.mm` and `.h` as defaults.

nikolaykasyanov commented:
@robertodr 👍 looks great.

It seems possible to make a test like existing "Reports code style violation as error", but such that does not specify `.m` as extension, thus making Danger to not emit an error. It could be called "Does not report errors in files with unwanted extensions" (not the best wording, though).

Another crucial test I'd expect is to verify that `ignore_file_patterns` are indeed respected when some of the ignored files have extensions matching `file_extensions` (for example, we ignore `generated/` directory that contains `.m` file and we pass `file_extensions: [ '.m' ]`, in this case `generated/file.m` should indeed be ignored). This can be achieved by adding explicit `file_extensions` argument to the [invocation](https://github.com/flix-tech/danger-code_style_validation/blob/a9dc5f4a37e4684db7f193e84401dc65c64ba859/spec/code_style_validation_spec.rb#L55).

nikolaykasyanov commented:
I could assist you with mentioned tests tomorrow if needed.

robertodr commented:
That would be great. I might not have time to look at it tomorrow, though.

nikolaykasyanov commented:
@robertodr could you please [provide](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) me a push access to this Pull Request so I can play with tests?

nikolaykasyanov commented:
@robertodr nevermind, it seems that checkbox was already enabled.

nikolaykasyanov commented on lib/code_style_validation/plugin.rb:
```diff
@@ -38,19 +44,31 @@ def check(config = {})
+      message = ''
+      unless offending_files.empty?
+        message = 'Code style violations detected in the following files:' + "\n"  
+	offending_files.each do |file_name|
```
You seem to be using tabs here. Please use spaces.

nikolaykasyanov replied:
It could be helpful to configure Rubocop support in your editor. This repo doesn't have Rubocop's config yet, so it'll report some false positives, but it will also detect things like inconsistent indentation and trailing spaces.

nikolaykasyanov commented on lib/code_style_validation/plugin.rb:
```diff
@@ -38,19 +44,31 @@ def check(config = {})
+
+      message = ''
+      unless offending_files.empty?
+        message = 'Code style violations detected in the following files:' + "\n"  
```
Trailing space.

robertodr replied:
Ooops... Sorry. I haven't yet reconfigured my Neovim properly after reinstalling the OS.

nikolaykasyanov replied:
No worries!

nikolaykasyanov commented:
My test fails on Travis, looking into it.

robertodr commented:
My neovim configuration is back on track! I have fixed indentation inconsistencies and fixed #9 
I also had a crazy idea, that I don't know how to implement. Usually, one PR might touch more than one or two files in many places. The patches might become quite long and the comment from the bot might easily take up a large portion of the screen and obfuscate comments from reviewers. Here comes the crazy idea, would it be possible to have the bot write a Gist with the patch? The comment from the bot would then:
1. List the offending files.
2. Suggest actions to correct the errors.
3. Point users to the Gist with the patch.
What do you think?

nikolaykasyanov commented:
@robertodr thanks! I'll try this out in our internal environment for a couple of days before merging.

Don't you mind if I squash commits upon merging?

robertodr commented:
Just test for as long as needed. No problem with squashing, as long as it preserves authorship assignment. What do you think about the crazy idea? Does it make sense? Is it doable?

nikolaykasyanov commented:
@robertodr sorry, I didn't notice your big comment.

The idea about Gist sounds reasonable. The only concern is that it introduces a separate code path for open source projects (because posting diffs to Gist most certainly won't work for proprietary/closed source projects).

I wonder if GitHub (GitLab) supports some kind of folding, that might solve the problem of big patches cluttering the review.

robertodr commented:
Oh! Yes, of course. I had not thought about that. Well, this PR is complete. Anything else can be part of subsequent PRs. Would you agree to tag the code and mint a release when the PR is merged?

robertodr commented:
Thanks! It's been a pleasure collaborating with you on this.
